File: programs/pulsecast/src/state/round.rs
use anchor_lang::prelude::*;

#[derive(AnchorSerialize, AnchorDeserialize, Clone, Copy, Debug, Eq, InitSpace, PartialEq)]
pub enum RoundStatus {
    Scheduled,
    Betting,
    Watching,
    Resolved,
    Settled,
    Cancelled,
}

#[account]
#[derive(InitSpace)]
pub struct Round {
    pub id: u64,
    pub status: RoundStatus,
    pub open_at: i64,
    pub lock_at: i64,
    pub resolve_at: i64,
    pub start_price: i64,
    pub start_publish_time: i64,
    pub actual_price: i64,
    pub actual_publish_time: i64,
    pub total_pool: u64,
    pub total_score: u64,
    pub protocol_fee: u64,
    pub prediction_count: u16,
    pub scored_count: u16,
    pub entry_amount: u64,
    pub fee_bps: u16,
    pub max_error_bps: u16,
    pub bump: u8,
}

File: packages/shared/src/round.ts
import { z } from "zod";

export const publicRoundSchema = z.object({
  account: z.string().min(32).max(44),
  actualPrice: z.string().regex(/^-?\d+$/),
  actualPublishTime: z.number().int(),
  entryAmount: z.string().regex(/^\d+$/),
  feeBps: z.number().int().min(0).max(10_000),
  id: z.string().regex(/^\d+$/),
  lockAt: z.number().int(),
  maxErrorBps: z.number().int().min(0).max(10_000),
  openAt: z.number().int(),
  predictionCount: z.number().int().nonnegative(),
  resolveAt: z.number().int(),
  schemaVersion: z.enum(["current", "legacy-without-claims"]),
  startPrice: z.string().regex(/^-?\d+$/),
  startPublishTime: z.number().int(),
  status: z.enum(["scheduled", "betting", "watching", "resolved", "settled", "cancelled"]),
  totalPool: z.string().regex(/^\d+$/),
});

export const roundIndexSchema = z.object({
  rounds: z.array(publicRoundSchema).max(100),
  serverTimeMs: z.number().int().nonnegative(),
  source: z.literal("solana-devnet"),
});

export type PublicRound = z.infer<typeof publicRoundSchema>;
export type RoundIndex = z.infer<typeof roundIndexSchema>;
